make bindings generation more robust

vrothberg opened the pull request:
The Go gods did not shine upon us trying to understand what's going on
in #9000.  The symptom is that `go generate` did not add required
imports to a generated file, ultimately breaking subsequent compilation.

While it still remains unclear *why* Go is behaving like that, the
symptom disappears when `go generate` runs in module mode; that is
without `-mod=vendor` and without `GO111MODULE=off`.  This was
reproducible on two separate machines (Ubuntu and Fedora).

Also, when facing an unset GOPATH, set it to Go's default (i.e.,
$HOME/go) and make sure that GOBIN is in PATH since `goimports`
is required by `go generate`.

Fixes: #9000
Signed-off-by: Valentin Rothberg <rothberg@redhat.com>

@baude @rhatdan @edsantiago PTAL

Luap99 commented on Makefile:
```diff
@@ -98,6 +98,8 @@ ifeq ($(GOBIN),)
 GOBIN := $(FIRST_GOPATH)/bin
 endif
 
+export PATH := $(PATH):$(GOPATH)
```
I think this should be GOBIN and not GOPATH

vrothberg replied:
Good catch, thanks!

TomSweeneyRedHat commented:
LGTM

edsantiago commented:
I still can't get this to work on my f33 laptop. After the `go generate ./pkg/bindings/...` step, my tree is left in a bad state:
```diff
index 5aec4b479..46a57eae6 100644
--- a/pkg/bindings/play/types_kube_options.go
+++ b/pkg/bindings/play/types_kube_options.go
@@ -6,8 +6,8 @@ import (
        "strconv"
        "strings"

-       jsoniter "github.com/json-iterator/go"
-       "github.com/pkg/errors"
+       jsoniter "github.com/containers/libpod/src/github.com/json-iterator/go"
+       "github.com/containers/libpod/src/github.com/pkg/errors"
 )

 /*
```
Easily fixed via `git reset --hard`, but I have no idea how to debug this; and I suspect that if I'm running into this, others will too. I'm really uncomfortable with this `bindings` step running as part of a regular `make`.

edsantiago commented:
I still can't get this to work on my f33 laptop. After the `go generate ./pkg/bindings/...` step, my tree is left in a bad state:
```diff
index 5aec4b479..46a57eae6 100644
--- a/pkg/bindings/play/types_kube_options.go
+++ b/pkg/bindings/play/types_kube_options.go
@@ -6,8 +6,8 @@ import (
        "strconv"
        "strings"

-       jsoniter "github.com/json-iterator/go"
-       "github.com/pkg/errors"
+       jsoniter "github.com/containers/libpod/src/github.com/json-iterator/go"
+       "github.com/containers/libpod/src/github.com/pkg/errors"
 )

 /*
```
Easily fixed via `git reset --hard`, but I have no idea how to debug this; and I suspect that if I'm running into this, others will too. I'm really uncomfortable with this `bindings` step running as part of a regular `make`.

vrothberg commented:
@edsantiago thanks for checking!  I am not yet happy as well but suggest to go step by step.  Maybe we find a better way in the (hopefully) near future.

Luap99 commented:
I also believe we should not run this for make or make binaries. All `pkg/bindings` files are committed. You only need to run `make .generate-bindings` when you changed something in `pkg/bindings/*/types.go`. I think this can be up to the individual developer in that case. It is also easy to enforce in CI, e.g. run generate bindings and check if the tree is still clean.

mheon commented:
My ideal situation would be that this `make` target is triggered by changes in those files, and otherwise does not fire.

vrothberg commented:
> My ideal situation would be that this `make` target is triggered by changes in those files, and otherwise does not fire.

In this case, we have to add the `.generate-bindings` file to Git.  But notice that we're a distributed team (different time zones) and that Make works on timestamps. So every time we update the file (and re-generate the bindings), there may be some false positives for a day.

vrothberg commented:
> I also believe we should not run this for make or make binaries. All `pkg/bindings` files are committed. You only need to run `make .generate-bindings` when you changed something in `pkg/bindings/*/types.go`. I think this can be up to the individual developer in that case. It is also easy to enforce in CI, e.g. run generate bindings and check if the tree is still clean.

I don't agree. History shows that if things are up to individual developers, things go wrong. I don't want to go back to the pre `make vendor` days.

We need to make sure that the generated code is _always_ in sync.

vrothberg commented:
> > I also believe we should not run this for make or make binaries. All `pkg/bindings` files are committed. You only need to run `make .generate-bindings` when you changed something in `pkg/bindings/*/types.go`. I think this can be up to the individual developer in that case. It is also easy to enforce in CI, e.g. run generate bindings and check if the tree is still clean.
> 
> I don't agree. History shows that if things are up to individual developers, things go wrong. I don't want to go back to the pre `make vendor` days.
> 
> We need to make sure that the generated code is _always_ in sync.

I just realize now that this is not mutually exclusive.  As @Luap99 said, we can easily enforce it in the CI.  That may be the smoothest way forward.

vrothberg commented:
/hold

vrothberg commented:
Alright, the new CI job is working: https://github.com/containers/podman/pull/9020/checks?check_run_id=1729142191

vrothberg commented:
Updated as discussed:
* `make generate-bindings` is explicit now
* Added a CI check similar to `make vendor` and integrated that together with the vendor checks into "code consistency" task.

@edsantiago @Luap99 @mheon PTAL

vrothberg commented:
@cevich PTAL

edsantiago commented on Makefile:
```diff
@@ -459,12 +460,10 @@ podman-remote-%-release:
 
-BINDINGS_SOURCE = $(wildcard pkg/bindings/**/types.go)
-.generate-bindings: $(BINDINGS_SOURCE)
+generate-bindings:
```
I think you need `.PHONY: generage-bindings` right above this

vrothberg replied:
Thanks for checking!

mheon commented:
LGTM once Ed's comment is addressed. Hoping we can get this landed soon, having to fight bindings generation is not fun.

edsantiago commented on Makefile:
```diff
@@ -463,12 +464,11 @@ podman-remote-%-release:
 
-BINDINGS_SOURCE = $(wildcard pkg/bindings/**/types.go)
-.generate-bindings: $(BINDINGS_SOURCE)
+.PHONY: geenrate-bindings
```
Aw, man, sorry to do this, but that's a typo

vrothberg replied:
Too many things in parallel ... sorry for my sloppiness

edsantiago commented:
Tests green.

/lgtm

There's a prior hold, remove it at will

vrothberg commented:
/hold cancel

vrothberg commented:
Thanks everybody!  I think we made some good improvements

Luap99 commented:
I think we should backport this, right?
